woocommerce/woocommerce
# Add changelogger validation.

### All Submissions:






### Changes proposed in this Pull Request:

This PR adapts the [script from Jetpack](https://github.com/Automattic/jetpack/blob/master/tools/check-changelogger-use.php) to check that any PR affecting a project configured for the changelogger has a changelog file present.

Instead of relying on the directory structure, projects are read from `workspace.json`. Any project with a `composer.json` file that includes the Jetpack Changelogger is considered a Changelogger project and therefore requires a changelog entry.

### How to test the changes in this Pull Request:

1. Create a new branch with changes to a project having changelogger enabled (such as `plugins/woocommerce`), but do not run the changelogger or add a changelog file.
2. Switch to this branch.
3. Run `php tools/monorepo/check-changelogger-use.php -h`. Verify that you see usage instructions and then it exits.
4. Run `php tools/monorepo/check-changelogger-use.php` (with no args). Verify that you see a message that base and head args are required, that usage instructions are provided, and then it exits.
5. Run `php tools/monorepo/check-changelogger-use.php refs/heads/trunk refs/heads/BRANCH && echo 'success'`. Replace "BRANCH" with your branch name. Verify that "success" **is not** echoed.
6. Run `php tools/monorepo/check-changelogger-use.php --debug refs/heads/trunk refs/heads/BRANCH`, again, replace BRANCH with your branch name. Verify that you see debug information, including "Project plugins/woocommerce is being changed, but no change file in plugins/woocommerce/plugins/woocommerce/changelog/ is touched". Verify that information about how to add a changelog file is displayed.
7. Add a changelog file to your branch by running the changelogger (first switch to your branch) with the provided command. Be sure to commit the change file.
8. Switch back to this branch.
9. Run `php tools/monorepo/check-changelogger-use.php refs/heads/trunk refs/heads/BRANCH && echo 'success'`. Replace "BRANCH" with your branch name. Verify that "success" **is** echoed.
10.  Run `php tools/monorepo/check-changelogger-use.php --debug refs/heads/trunk refs/heads/BRANCH`, again, replace BRANCH with your branch name. Verify that you see debug information, but you don't see any errors or information about how to run the changelogger.
11. Create a new branch with changes that don't affect a project with changelogger enabled (such as touching a file in the root of the repo, or making changes to a non-changelogger project).
12. Switch back to this branch.
13. Run `php tools/monorepo/check-changelogger-use.php refs/heads/trunk refs/heads/BRANCH && echo 'success'`. You should see success as there aren't any changes affecting a changelogger project.
14. Optional: Test the other flags of the changelogger project. For example, passing `--path=plugins/woocommerce-admin` should only check the woocommerce-admin file for changes, and not be affected by changes outside of that path.
15. Fork this repository.
16. Merge this PR into `trunk` on your forked repository.
17. Create a PR affecting a plugin with changelogger enabled, but do not run the changelogger tool.
18. Verify that the github action for checking for the changelogger file runs and fails.
19. Update the PR to add the changelogger file by running the changelogger tool.
20. Verify that the github action for checking for the changelogger file runs and passes.
21. Create a PR affecting a plugin without changelogger enabled. Verify the github action runs and passes even without a changelog file (as one is not needed).

### Other information:




### Changelog entry

No changelog entry required.

### FOR PR REVIEWER ONLY:
